docs/man/scope.1
.\" Automatically generated by Pandoc 3.1.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "TRACKER-SCOPE" "1" "2023/10" "" "Tracker Scope Flag Manual"
.hy
.SS NAME
.PP
tracker \f[B]--scope\f[R] - Select the scope for tracing events
.SS SYNOPSIS
.PP
tracker \f[B]--scope\f[R]
[<[uid|pid][=|!=|<|>|<=|>=]value1(,value2\&...)> |
<[mntns|pidns|tree][=|!=]value1(,value2\&...)> |
<[uts|comm|container|[executable|exec|binary|bin]][=|!=]value1(,value2\&...)>]
| <not-container> | <container[=|!=]value> | <[container|pid]=new> |
<follow>] \&...
.SS DESCRIPTION
.PP
The \f[B]--scope\f[R] flag allows you to select the scope for tracing
events by defining filters.
.SS FILTER EXPRESSION
.PP
Filter expressions can be defined to operate on scope options or process
metadata.
Only events that match all filter expressions will be traced.
.PP
Multiple flags are combined with AND logic, while multiple values within
a single flag are combined with OR logic when using the equals operator
`='.
Multiple values can be specified using `,'.
.SS NUMERICAL EXPRESSION OPERATORS
.PP
The following numerical fields support the operators `=', `!=', `<',
`>', `<=', `>=':
.IP \[bu] 2
uid: Select events from specific user IDs.
.IP \[bu] 2
pid: Select events from specific process IDs.
.PP
The following numerical fields only support the operators `=' and `!=':
.IP \[bu] 2
mntns: Select events from specific mount namespace IDs.
.IP \[bu] 2
pidns: Select events from specific process namespace IDs.
.IP \[bu] 2
tree: Select events that descend from specific process IDs.
.PP
NOTE: Expressions containing `<' or `>' tokens must be escaped!
.SS STRING EXPRESSION OPERATORS
.PP
`=', `!='
.PP
Available for the following string fields:
.IP \[bu] 2
uts: Select events based on UTS (Unix Timesharing System) names.
.IP \[bu] 2
comm: Select events based on process command names.
.IP \[bu] 2
container: Select events from specific container IDs.
.IP \[bu] 2
executable: Select events based on the executable path.
.PP
Strings can be compared as a prefix if ending with `*', or as a suffix
if starting with `*'.
.PP
NOTE: Expressions containing `*' token must be escaped!
.SS BOOLEAN OPERATOR (PREPENDED)
.PP
`!'
.PP
Available for the following boolean field:
.IP \[bu] 2
container: Select events based on whether they originate from a
container or not.
.SS SPECIAL FILTERS
.PP
The following special filters can be used within the scope filter
expressions:
.IP \[bu] 2
new: Select newly created containers or process IDs.
.IP \[bu] 2
follow: Select events from the processes that match the criteria and
their descendants.
.SS EXAMPLES
.IP \[bu] 2
To trace only events from new processes, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope pid=new
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from pid 510 or pid 1709, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope pid=510,1709
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from pid 510 or pid 1709 (same as above), use the
following flag:
.RS 2
.IP
.nf
\f[C]
--scope p=510 --scope p=1709
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from newly created containers, use the following
flag:
.RS 2
.IP
.nf
\f[C]
--scope container=new
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from the container with ID `ab356bc4dd554', use the
following flag:
.RS 2
.IP
.nf
\f[C]
--scope container=ab356bc4dd554
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from containers, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope container
\f[R]
.fi
.RE
.IP \[bu] 2
To only trace events from containers (same as above), use the following
flag:
.RS 2
.IP
.nf
\f[C]
--scope c
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from the host, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope not-container
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from uid 0, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope uid=0
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from mntns id 4026531840, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope mntns=4026531840
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from pidns id not equal to 4026531836, use the
following flag:
.RS 2
.IP
.nf
\f[C]
--scope pidns!=4026531836
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events that descend from the process with pid 476165, use
the following flag:
.RS 2
.IP
.nf
\f[C]
--scope tree=476165
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events if they do not descend from the process with pid
5023, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope tree!=5023
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events if they descend from 3213 or 5200, but not 3215,
use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope tree=3213,5200 --scope tree!=3215
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from uids greater than 0, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope \[aq]uid>0\[aq]
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from pids between 0 and 1000, use the following
flag:
.RS 2
.IP
.nf
\f[C]
--scope \[aq]pid>0\[aq] --scope \[aq]pid<1000\[aq]
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from uids greater than 0 but not 1000, use the
following flag:
.RS 2
.IP
.nf
\f[C]
--scope \[aq]u>0\[aq] --scope u!=1000
\f[R]
.fi
.RE
.IP \[bu] 2
To exclude events from uts name `ab356bc4dd554', use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope uts!=ab356bc4dd554
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from the `ls' command, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope comm=ls
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from the `/usr/bin/ls' executable, use the
executable flag (or the binary alias):
.RS 2
.IP
.nf
\f[C]
--scope executable=/usr/bin/ls
\f[R]
.fi
.IP
.nf
\f[C]
--scope binary=/usr/bin/ls
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from the `/usr/bin/ls' executable in the host mount
namespace, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope executable=host:/usr/bin/ls
\f[R]
.fi
.RE
.IP \[bu] 2
To trace only events from the `/usr/bin/ls' executable in the 4026532448
mount namespace, use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope executable=4026532448:/usr/bin/ls
\f[R]
.fi
.RE
.IP \[bu] 2
To trace all events that originated from `bash' or from one of the
processes spawned by `bash', use the following flag:
.RS 2
.IP
.nf
\f[C]
--scope comm=bash --scope follow
\f[R]
.fi
.RE


pkg/cmd/flags/scope.go
package flags

import (
	"strings"

	"github.com/khulnasoft-lab/tracker/pkg/errfmt"
)

// PolicyScopeMap maps policy id to its pre-parsed scope flag fields
type PolicyScopeMap map[int]policyScopes

// policyScopes holds pre-parsed scope flag fields of one policy
type policyScopes struct {
	policyName string
	scopeFlags []scopeFlag
}

// scopeFlag holds pre-parsed scope flag fields
type scopeFlag struct {
	full              string
	scopeName         string
	operator          string
	values            string
	operatorAndValues string
}

func PrepareScopeMapFromFlags(filtersArr []string) (PolicyScopeMap, error) {
	// parse and store scope flags
	var filterFlags []scopeFlag
	for _, filter := range filtersArr {
		parsed, err := parseScopeFlag(filter)
		if err != nil {
			return nil, err
		}

		filterFlags = append(filterFlags, parsed)
	}

	filterMap := make(PolicyScopeMap)
	filterMap[0] = policyScopes{scopeFlags: filterFlags}

	return filterMap, nil
}

// parseScopeFlag parses a scope flag and returns a scopeFlag struct with
// pre-parsed fields, or an error if the flag is invalid.
func parseScopeFlag(flag string) (scopeFlag, error) {
	if flag == "" {
		return scopeFlag{}, errfmt.WrapError(InvalidFlagEmpty())
	}

	// get first idx of any expression operator (=, !=, <, >, <=, >=)
	operatorIdx := strings.IndexAny(flag, "=!<>")

	//
	// without expression operator
	//

	if operatorIdx == -1 { // no expression operator
		if hasLeadingOrTrailingWhitespace(flag) {
			return scopeFlag{}, errfmt.WrapError(InvalidFilterFlagFormat(flag))
		}

		// unset flag
		unsetPrefix := "not-"
		if strings.HasPrefix(flag, unsetPrefix) {
			if len(flag) == len(unsetPrefix) {
				return scopeFlag{}, errfmt.WrapError(InvalidFilterFlagFormat(flag))
			}

			name := flag[len(unsetPrefix):]
			if hasLeadingOrTrailingWhitespace(name) {
				return scopeFlag{}, errfmt.WrapError(InvalidFilterFlagFormat(flag))
			}

			return scopeFlag{
				full:      flag,
				scopeName: name,
				operator:  unsetPrefix[:len(unsetPrefix)-1],
			}, nil
		}

		// set flag
		return scopeFlag{
			full:      flag,
			scopeName: flag,
		}, nil
	}

	//
	// with expression operator
	//

	// validate scope name
	scopeName := flag[:operatorIdx]
	if hasLeadingOrTrailingWhitespace(scopeName) {
		return scopeFlag{}, errfmt.WrapError(InvalidFilterFlagFormat(flag))
	}

	// validate operator and values
	opAndValParts, err := getOperatorAndValuesParts(flag, operatorIdx)
	if err != nil {
		return scopeFlag{}, errfmt.WrapError(err)
	}

	return scopeFlag{
		full:              flag,                            // "executable=host:/usr/bin/ls"
		scopeName:         scopeName,                       // "executable"
		operator:          opAndValParts.operator,          // "="
		values:            opAndValParts.values,            // "host:/usr/bin/ls"
		operatorAndValues: opAndValParts.operatorAndValues, // "=host:/usr/bin/ls"
	}, nil
}
